### PR TITLE
create v2 version of MentionsCreationStateMachine to support custom chooser view

### PR DESCRIPTION
**Pull request by li-jychen**

### CODE REVIEW

File: Hakawai/Mentions/_HKWMentionsCreationStateMachine.h [hunk 1]
```diff
@@ -211,7 +192,11 @@
 /*!
  Handles the selection from the user.
  Needs to be public for integration between Hakawai and HotPot.
+ // TODO: remove indexPath as it's no longer needed for tracking
```
Comment (bkoatz) [inline Hakawai/Mentions/_HKWMentionsCreationStateMachine.h:195]: is there a JIRA for this/can you add one

File: HakawaiDemo/HakawaiDemo/MentionsDemoViewController.m [hunk 2]
```diff
@@ -53,10 +53,6 @@ - (void)viewDidLoad {
                                                                            controlCharacters:controlCharacters
                                                                                 searchLength:3];
 
-        // NOTE: If you want to see an example of a custom chooser, uncomment the following line.
-//        mentionsPlugin.chooserViewClass = [SimpleChooserView class];
```
Comment (bkoatz) [inline HakawaiDemo/HakawaiDemo/MentionsDemoViewController.m:57]: should we just add a customerChooserView option for tests? this should still be tested right? maybe add this to a hammer-testing jira?

File: HakawaiTests/HKWMentionsCreationStateMachineTest.m [hunk 3]
```diff
@@ -54,48 +70,115 @@ @interface HKWMentionsPlugin () <HKWMentionsDefaultChooserViewDelegate>
...
-        expect(mentionsPlugin.creationStateMachine.entityArray.count).to.equal(0);
+    NSArray *entityArray;
+    if (HKWTextView.enableMentionsCreationStateMachineV2) {
+        HKWMentionsCreationStateMachineV2 *sm2 = (HKWMentionsCreationStateMachineV2 *)mentionsPlugin.creationStateMachine;
+        entityArray = sm2.dataProvider.entityArray;
+    } else {
+        HKWMentionsCreationStateMachineV1 *sm1 = (HKWMentionsCreationStateMachineV1 *)mentionsPlugin.creationStateMachine;
+        entityArray = sm1.entityArray;
+    }
+        expect(entityArray.count).to.equal(0);
```
Comment (bkoatz) [inline HakawaiTests/HKWMentionsCreationStateMachineTest.m:81]: fix tabbing, same below/throughout

File: HakawaiTests/HKWMentionsCreationStateMachineTest.m [hunk 4]
```diff
@@ -20,20 +20,36 @@
...
 @end
 
 @interface HKWMentionsPlugin () <HKWMentionsDefaultChooserViewDelegate>
 
-@property (nonatomic, strong) HKWMentionsCreationStateMachine *creationStateMachine;
+@property (nonatomic, strong) id<HKWMentionsCreationStateMachine> creationStateMachine;
 
 @end
 
+
```
Comment (bkoatz) [inline HakawaiTests/HKWMentionsCreationStateMachineTest.m:52]: remove

File: Hakawai/Mentions/HKWMentionDataProvider.m [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+#import "HKWMentionDataProvider.h"
+#import "_HKWMentionsPrivateConstants.h"
+
+/*!
+ States for the network state subsidiary state machine.
+ */
+typedef NS_ENUM(NSInteger, HKWMentionsCreationNetworkState) {
+    // The mentions creation system is ready: data has been populated, and the system is waiting for the user to type
+    //  a character, select a mention, or perform another action.
+    HKWMentionsCreationNetworkStateReady = 0,
```
[the remark this answers is not in the document]
Comment (li-jychen) [reply Hakawai/Mentions/HKWMentionDataProvider.m:18]: dataProvider won't get called when it's network state quiescent. so it makes sense to remove

File: Hakawai/Mentions/HKWMentionDataProvider.m [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+        }
+    }
+}
+
+- (void)sendQueryWithKeyString:(nonnull NSString *)string
+                    searchType:(HKWMentionsSearchType)type
+                  isWhitespace:(BOOL)isWhitespace
+              controlCharacter:(unichar)character {
+    [self activateCooldownTimer];
+    NSLog(@"fire:%@",string);
```
[the remark this answers is not in the document]
Comment (li-jychen) [reply Hakawai/Mentions/HKWMentionDataProvider.m:108]: remove what? the log? it's originally there in mentionstatemachine

File: Hakawai/Mentions/HKWMentionsCreationStateMachineV1.m [hunk 7]
```diff
@@ -0,0 +1,1171 @@
+//
```
[the remark this answers is not in the document]
Comment (li-jychen) [reply Hakawai/Mentions/HKWMentionsCreationStateMachineV1.m:1]: Yes

File: Hakawai/Mentions/HKWMentionsCreationStateMachineV2.m [hunk 8]
```diff
@@ -0,0 +1,708 @@
...
+        chooserFrame.size.height -= (self.chooserViewEdgeInsets.top + self.chooserViewEdgeInsets.bottom);
+        chooserFrame.origin.x += self.chooserViewEdgeInsets.left;
+        chooserFrame.origin.y += self.chooserViewEdgeInsets.top;
+    }
+    else {
+        chooserFrame = CGRectNull;
+    }
+    return chooserFrame;
+}
+
```
Comment (bkoatz) [inline Hakawai/Mentions/HKWMentionsCreationStateMachineV2.m:408]: extra space

File: Hakawai/Mentions/HKWMentionsCreationStateMachineV2.m [hunk 9]
```diff
@@ -0,0 +1,707 @@
...
+    NSAssert((delegate.shouldContinueSearchingAfterEmptyResults && self.resultsState == HKWMentionsCreationResultsStateAwaitingFirstResult)
+             || self.resultsState == HKWMentionsCreationResultsStateCreatingMentionWithResults
+             || self.resultsState == HKWMentionsCreationResultsStateNoResultsWithoutWhitespace
+             || self.searchType == HKWMentionsSearchTypeInitial,
+             @"Logic error in dataReturnedForResults:...; resultsState is inconsistent. Got %@, which is invalid.",
+             nameForResultsState(self.resultsState));
+    self.resultsState = HKWMentionsCreationResultsStateNoResultsWithoutWhitespace;
+}
+
+
```
Comment (bkoatz) [inline Hakawai/Mentions/HKWMentionsCreationStateMachineV2.m:557]: extra space

**Pull request merged**